File: apps/api/src/modules/kafka/kafka-producer.service.ts
import { Injectable, Logger, OnModuleDestroy, OnModuleInit } from '@nestjs/common';
import { Kafka, logLevel, type Consumer, type Producer } from 'kafkajs';
import { KafkaConfigService } from './kafka.config';
import type { DualVerifyJobMessage } from '../dual-verify-kafka/dual-verify-kafka.types';

export type JobMessageHandler = (message: DualVerifyJobMessage) => Promise<void>;

@Injectable()
export class KafkaProducerService implements OnModuleInit, OnModuleDestroy {
  private readonly logger = new Logger(KafkaProducerService.name);
  private kafka: Kafka | null = null;
  private consumerKafka: Kafka | null = null;
  private producer: Producer | null = null;
  private consumer: Consumer | null = null;
  private handler: JobMessageHandler | null = null;
  private consuming = false;

  constructor(private readonly kafkaConfig: KafkaConfigService) {}

  async onModuleInit(): Promise<void> {
    if (!this.kafkaConfig.isEnabled()) {
      this.logger.log('Kafka disabled — using in-process local queue');
      return;
    }

    this.kafka = new Kafka({
      clientId: this.kafkaConfig.getClientId(),
      brokers: this.kafkaConfig.getBrokers(),
      ssl: true,
      sasl: {
        mechanism: 'plain',
        username: '$ConnectionString',
        password: this.kafkaConfig.getProducerPassword(),
      },
      logLevel: logLevel.WARN,
    });

    this.producer = this.kafka.producer();
    await this.producer.connect();
    this.logger.log(
      `Kafka producer connected (${this.kafkaConfig.getBrokers().join(', ')})`,
    );
  }

  async onModuleDestroy(): Promise<void> {
    this.consuming = false;
    if (this.consumer) {
      await this.consumer.disconnect().catch(() => undefined);
    }
    if (this.producer) {
      await this.producer.disconnect().catch(() => undefined);
    }
    this.consumerKafka = null;
    this.kafka = null;
  }

  getTransport(): 'kafka' | 'local' {
    return this.kafkaConfig.getTransportMode();
  }

  isKafkaReady(): boolean {
    return Boolean(this.producer);
  }

  /** Publish one dual-verify job message */
  async publishJob(message: DualVerifyJobMessage): Promise<void> {
    if (!this.producer) {
      throw new Error('Kafka producer not connected');
    }

    const topic = this.kafkaConfig.getTopicJobs();
    const key = `${message.sessionId}:${message.pointId}`;

    await this.producer.send({
      topic,
      messages: [
        {
          key,
          value: JSON.stringify(message),
          headers: {
            schemaVersion: message.schemaVersion,
            correlationId: message.correlationId,
            attempt: String(message.attempt),
          },
        },
      ],
    });

    this.logger.debug(`Published job ${message.jobId} → ${topic} (${key})`);
  }

  /** Publish to retry or DLQ topic (worker-send policy when configured) */
  async publishToTopic(
    topic: 'retry' | 'dlq' | 'results',
    message: DualVerifyJobMessage | Record<string, unknown>,
  ): Promise<void> {
    if (!this.kafka && !this.producer) return;

    const topicName =
      topic === 'retry'
        ? this.kafkaConfig.getTopicRetry()
        : topic === 'dlq'
          ? this.kafkaConfig.getTopicDlq()
          : this.kafkaConfig.getTopicResults();

    const payload =
      'sessionId' in message && 'pointId' in message
        ? {
            key: `${message.sessionId}:${message.pointId}`,
            value: JSON.stringify(message),
          }
        : { value: JSON.stringify(message) };

    const useWorkerSend =
      (topic === 'retry' || topic === 'dlq') &&
      this.kafkaConfig.getWorkerSendPassword() !==
        this.kafkaConfig.getProducerPassword();

    if (useWorkerSend && this.kafka) {
      const client = new Kafka({
        clientId: `${this.kafkaConfig.getClientId()}-worker-send`,
        brokers: this.kafkaConfig.getBrokers(),
        ssl: true,
        sasl: {
          mechanism: 'plain',
          username: '$ConnectionString',
          password: this.kafkaConfig.getWorkerSendPassword(),
        },
        logLevel: logLevel.WARN,
      });
      const tempProducer = client.producer();
      await tempProducer.connect();
      await tempProducer.send({ topic: topicName, messages: [payload] });
      await tempProducer.disconnect();
      return;
    }

    if (!this.producer) return;
    await this.producer.send({ topic: topicName, messages: [payload] });
  }

  /** Start Kafka consumer loop (listen-only connection string) */
  async startConsumer(handler: JobMessageHandler): Promise<void> {
    if (this.consuming && this.consumer) {
      this.handler = handler;
      return;
    }

    if (this.consumer) {
      await this.consumer.disconnect().catch(() => undefined);
      this.consumer = null;
      this.consuming = false;
    }

    this.handler = handler;

    this.consumerKafka = new Kafka({
      clientId: `${this.kafkaConfig.getClientId()}-consumer`,
      brokers: this.kafkaConfig.getBrokers(),
      ssl: true,
      sasl: {
        mechanism: 'plain',
        username: '$ConnectionString',
        password: this.kafkaConfig.getConsumerPassword(),
      },
      logLevel: logLevel.WARN,
    });

    this.consumer = this.consumerKafka.consumer({
      groupId: this.kafkaConfig.getConsumerGroup(),
      sessionTimeout: 30_000,
      heartbeatInterval: 3_000,
      maxWaitTimeInMs: 5_000,
    });

    await this.consumer.connect();
    await this.consumer.subscribe({
      topic: this.kafkaConfig.getTopicJobs(),
      fromBeginning: false,
    });

    this.consuming = true;
    this.logger.log(
      `Kafka consumer started (group=${this.kafkaConfig.getConsumerGroup()})`,
    );

    void this.consumer
      .run({
        autoCommit: false,
        eachMessage: async ({ message, partition, topic, heartbeat }) => {
          if (!this.handler || !message.value) return;

          this.logger.log(
            `Kafka message received partition=${partition} offset=${message.offset}`,
          );

          let job: DualVerifyJobMessage;
        try {
          job = JSON.parse(message.value.toString()) as DualVerifyJobMessage;
        } catch {
          this.logger.error('Invalid Kafka message JSON — skipping');
          if (this.consumer) {
            await this.consumer.commitOffsets([
              {
                topic,
                partition,
                offset: (BigInt(message.offset) + 1n).toString(),
              },
            ]);
          }
          return;
        }

        await heartbeat();
        await this.handler(job);

        if (this.consumer) {
          await this.consumer.commitOffsets([
            {
              topic,
              partition,
              offset: (BigInt(message.offset) + 1n).toString(),
            },
          ]);
        }
      },
    }).catch((err: unknown) => {
      this.consuming = false;
      this.logger.error(
        `Kafka consumer run loop stopped: ${err instanceof Error ? err.message : err}`,
      );
    });
  }
}

File: apps/reguliq-dotnet/src/Reguliq.Api/Infrastructure/KafkaProducerService.cs
using System.Text;
using System.Text.Json;
using Confluent.Kafka;
using Reguliq.Api.Models;

namespace Reguliq.Api.Infrastructure;

public delegate Task KafkaJobHandler(DualVerifyJobMessage message, CancellationToken ct);

/// <summary>Kafka producer + jobs-topic consumer (Azure Event Hubs SASL).</summary>
public class KafkaProducerService(
    KafkaConfig kafkaConfig,
    ILogger<KafkaProducerService> logger) : IAsyncDisposable
{
    private IProducer<string, string>? _producer;
    private IConsumer<string, string>? _consumer;
    private CancellationTokenSource? _consumerCts;
    private Task? _consumerTask;
    private KafkaJobHandler? _handler;

    public bool IsReady => _producer != null;

    public string GetTransport() => kafkaConfig.GetTransportMode();

    public async Task ConnectProducerAsync(CancellationToken ct = default)
    {
        if (!kafkaConfig.IsEnabled()) return;
        if (_producer != null) return;

        var config = BuildClientConfig(kafkaConfig.GetProducerPassword());
        _producer = new ProducerBuilder<string, string>(config).Build();
        logger.LogInformation("Kafka producer connected ({Brokers})", string.Join(", ", kafkaConfig.GetBrokers()));
        await Task.CompletedTask;
    }

    public async Task PublishJobAsync(DualVerifyJobMessage message, CancellationToken ct = default)
    {
        if (_producer == null)
            throw new InvalidOperationException("Kafka producer not connected");

        var topic = kafkaConfig.GetTopicJobs();
        var key = $"{message.SessionId}:{message.PointId}";
        var json = JsonSerializer.Serialize(message, JsonOptions);

        await _producer.ProduceAsync(topic, new Message<string, string> { Key = key, Value = json }, ct);
        logger.LogDebug("Published job {JobId} → {Topic} ({Key})", message.JobId, topic, key);
    }

    public async Task PublishToTopicAsync(
        string topicKind,
        object payload,
        CancellationToken ct = default)
    {
        if (_producer == null) return;

        var topicName = topicKind switch
        {
            "retry" => kafkaConfig.GetTopicRetry(),
            "dlq" => kafkaConfig.GetTopicDlq(),
            "results" => kafkaConfig.GetTopicResults(),
            _ => topicKind
        };

        var json = JsonSerializer.Serialize(payload, JsonOptions);
        string? key = null;
        if (payload is DualVerifyJobMessage job)
            key = $"{job.SessionId}:{job.PointId}";
        else if (payload is JsonElement el && el.TryGetProperty("sessionId", out var sid) && el.TryGetProperty("pointId", out var pid))
            key = $"{sid.GetString()}:{pid.GetString()}";

        var useWorkerSend = (topicKind is "retry" or "dlq")
            && kafkaConfig.GetWorkerSendPassword() != kafkaConfig.GetProducerPassword();

        if (useWorkerSend)
        {
            using var temp = new ProducerBuilder<string, string>(
                BuildClientConfig(kafkaConfig.GetWorkerSendPassword())).Build();
            await temp.ProduceAsync(topicName, new Message<string, string> { Key = key, Value = json }, ct);
            return;
        }

        await _producer.ProduceAsync(topicName, new Message<string, string> { Key = key, Value = json }, ct);
    }

    public async Task StartConsumerAsync(KafkaJobHandler handler, CancellationToken ct = default)
    {
        if (!kafkaConfig.IsEnabled()) return;
        _handler = handler;

        if (_consumer != null)
        {
            _handler = handler;
            return;
        }

        var config = BuildClientConfig(kafkaConfig.GetConsumerPassword());
        var consumerConfig = new ConsumerConfig(config)
        {
            GroupId = kafkaConfig.GetConsumerGroup(),
            AutoOffsetReset = AutoOffsetReset.Latest,
            EnableAutoCommit = false,
            SessionTimeoutMs = 30_000,
            MaxPollIntervalMs = 300_000
        };

        _consumer = new ConsumerBuilder<string, string>(consumerConfig).Build();
        _consumer.Subscribe(kafkaConfig.GetTopicJobs());
        _consumerCts = CancellationTokenSource.CreateLinkedTokenSource(ct);
        _consumerTask = Task.Run(() => ConsumeLoopAsync(_consumerCts.Token), _consumerCts.Token);

        logger.LogInformation(
            "Kafka consumer started (group={Group}, topic={Topic})",
            kafkaConfig.GetConsumerGroup(),
            kafkaConfig.GetTopicJobs());
        await Task.CompletedTask;
    }

    private async Task ConsumeLoopAsync(CancellationToken ct)
    {
        if (_consumer == null || _handler == null) return;

        while (!ct.IsCancellationRequested)
        {
            try
            {
                var cr = _consumer.Consume(ct);
                if (cr?.Message?.Value == null) continue;
                if (cr.Message.Key == "health-check") { _consumer.Commit(cr); continue; }

                DualVerifyJobMessage? job;
                try
                {
                    job = JsonSerializer.Deserialize<DualVerifyJobMessage>(cr.Message.Value, JsonOptions);
                }
                catch
                {
                    logger.LogError("Invalid Kafka message JSON — skipping");
                    _consumer.Commit(cr);
                    continue;
                }

                if (job == null) { _consumer.Commit(cr); continue; }

                await _handler(job, ct);
                _consumer.Commit(cr);
            }
            catch (ConsumeException ex) when (ex.Error.IsFatal)
            {
                logger.LogError(ex, "Kafka consumer fatal error");
                break;
            }
            catch (OperationCanceledException) when (ct.IsCancellationRequested)
            {
                break;
            }
            catch (Exception ex)
            {
                logger.LogError(ex, "Kafka consumer loop error");
                await Task.Delay(1000, ct);
            }
        }
    }

    private ClientConfig BuildClientConfig(string password) => new()
    {
        BootstrapServers = string.Join(",", kafkaConfig.GetBrokers()),
        ClientId = kafkaConfig.GetClientId(),
        SecurityProtocol = SecurityProtocol.SaslSsl,
        SaslMechanism = SaslMechanism.Plain,
        SaslUsername = "$ConnectionString",
        SaslPassword = password
    };

    private static readonly JsonSerializerOptions JsonOptions = new()
    {
        PropertyNamingPolicy = JsonNamingPolicy.CamelCase,
        PropertyNameCaseInsensitive = true
    };

    public async ValueTask DisposeAsync()
    {
        _consumerCts?.Cancel();
        if (_consumerTask != null)
        {
            try { await _consumerTask; }
            catch (OperationCanceledException) { /* expected */ }
        }
        _consumer?.Close();
        _consumer?.Dispose();
        if (_producer != null)
        {
            _producer.Flush(TimeSpan.FromSeconds(3));
            _producer.Dispose();
        }
    }
}
